// obyte/lib/client.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _wsclient = require('./wsclient');

var _wsclient2 = _interopRequireDefault(_wsclient);

var _utils = require('./utils');

var _constants = require('./constants');

var _internal = require('./internal');

var _api = require('./api.json');

var _api2 = _interopRequireDefault(_api);

var _apps = require('./apps.json');

var _apps2 = _interopRequireDefault(_apps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Client = function () {
  function Client() {
    var _this = this;

    var nodeAddress = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _constants.DEFAULT_NODE;
    var clientOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    (0, _classCallCheck3.default)(this, Client);

    var self = this;

    this.options = (typeof clientOptions === 'undefined' ? 'undefined' : (0, _typeof3.default)(clientOptions)) === 'object' ? clientOptions : { testnet: clientOptions };
    this.client = new _wsclient2.default(nodeAddress);
    this.cachedWitnesses = null;

    var requestAsync = function requestAsync(name, params) {
      return new _promise2.default(function (resolve, reject) {
        _this.client.request(name, params, function (err, result) {
          if (err) return reject(err);
          return resolve(result);
        });
      });
    };

    this.api = {};

    this.compose = {
      message: function () {
        var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(app, payload) {
          var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

          var isDefinitionRequired, conf, privKeyBuf, pubkey, definition, address, path, version, bJsonBased, witnesses, _ref2, _ref3, lightProps, objDefinition, bytePayment, customMessages, assetPayment, unit, author, assocSigningPaths, assocLengthsBySigningPaths, arrSigningPaths, j, headersCommission, payloadCommission, textToSign;

          return _regenerator2.default.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  isDefinitionRequired = false;
                  conf = (typeof options === 'undefined' ? 'undefined' : (0, _typeof3.default)(options)) === 'object' ? (0, _extends3.default)({}, self.options, options) : (0, _extends3.default)({}, self.options, { wif: options });
                  privKeyBuf = conf.privateKey || (0, _utils.fromWif)(conf.wif, conf.testnet).privateKey;
                  pubkey = (0, _internal.toPublicKey)(privKeyBuf);
                  definition = conf.definition || ['sig', { pubkey: pubkey }];
                  address = conf.address || (0, _utils.getChash160)(definition);
                  path = conf.path || 'r';
                  version = conf.testnet ? _constants.VERSION_TESTNET : _constants.VERSION;
                  bJsonBased = version !== _constants.VERSION_WITHOUT_TIMESTAMP;
                  _context.next = 11;
                  return self.getCachedWitnesses();

                case 11:
                  witnesses = _context.sent;
                  _context.next = 14;
                  return _promise2.default.all([self.api.getParentsAndLastBallAndWitnessListUnit({ witnesses: witnesses }), self.api.getDefinitionForAddress({ address: address })]);

                case 14:
                  _ref2 = _context.sent;
                  _ref3 = (0, _slicedToArray3.default)(_ref2, 2);
                  lightProps = _ref3[0];
                  objDefinition = _ref3[1];

                  if (!(!objDefinition.definition && objDefinition.is_stable)) {
                    _context.next = 22;
                    break;
                  }

                  isDefinitionRequired = true;
                  _context.next = 24;
                  break;

                case 22:
                  if (objDefinition.is_stable) {
                    _context.next = 24;
                    break;
                  }

                  throw new Error('Definition or definition change for address ' + address + ' is not stable yet');

                case 24:
                  if (!(objDefinition.definition_chash !== (0, _utils.getChash160)(definition))) {
                    _context.next = 26;
                    break;
                  }

                  throw new Error('Definition chash of address doesn\'t match the definition chash provided');

                case 26:
                  _context.next = 28;
                  return (0, _internal.createPaymentMessage)(self, null, app !== 'payment' || payload.asset ? [] : payload.outputs, address);

                case 28:
                  bytePayment = _context.sent;
                  customMessages = [bytePayment];

                  if (!(app === 'payment')) {
                    _context.next = 38;
                    break;
                  }

                  if (!payload.asset) {
                    _context.next = 36;
                    break;
                  }

                  _context.next = 34;
                  return (0, _internal.createPaymentMessage)(self, payload.asset, payload.outputs, address);

                case 34:
                  assetPayment = _context.sent;

                  customMessages.push(assetPayment);

                case 36:
                  _context.next = 39;
                  break;

                case 38:
                  customMessages.push({
                    app: app,
                    payload_hash: (0, _internal.getBase64Hash)(payload, bJsonBased),
                    payload_location: 'inline',
                    payload: payload
                  });

                case 39:
                  unit = {
                    version: conf.testnet ? _constants.VERSION_TESTNET : _constants.VERSION,
                    alt: conf.testnet ? _constants.ALT_TESTNET : _constants.ALT,
                    messages: [].concat(customMessages),
                    authors: [],
                    parent_units: lightProps.parent_units,
                    last_ball: lightProps.last_stable_mc_ball,
                    last_ball_unit: lightProps.last_stable_mc_ball_unit,
                    witness_list_unit: lightProps.witness_list_unit,
                    timestamp: Math.round(Date.now() / 1000)
                  };
                  author = { address: address, authentifiers: {} };

                  if (isDefinitionRequired) {
                    author.definition = definition;
                  }

                  assocSigningPaths = {};
                  assocLengthsBySigningPaths = { r: 88 };
                  arrSigningPaths = (0, _keys2.default)(assocLengthsBySigningPaths);

                  assocSigningPaths[address] = arrSigningPaths;
                  for (j = 0; j < arrSigningPaths.length; j += 1) {
                    author.authentifiers[arrSigningPaths[j]] = (0, _internal.repeatString)('-', assocLengthsBySigningPaths[arrSigningPaths[j]]);
                  }
                  unit.authors.push(author);

                  headersCommission = (0, _internal.getHeadersSize)(unit);
                  payloadCommission = (0, _internal.getTotalPayloadSize)(unit);


                  customMessages[0].payload.outputs[0].amount -= headersCommission + payloadCommission;
                  customMessages[0].payload.outputs.sort(_internal.sortOutputs);
                  customMessages[0].payload_hash = (0, _internal.getBase64Hash)(customMessages[0].payload, bJsonBased);

                  if (payload.asset) {
                    customMessages[1].payload.outputs.sort(_internal.sortOutputs);
                    customMessages[1].payload_hash = (0, _internal.getBase64Hash)(customMessages[1].payload, bJsonBased);
                  }

                  unit.headers_commission = headersCommission;
                  unit.payload_commission = payloadCommission;

                  textToSign = (0, _internal.getUnitHashToSign)(unit);

                  unit.authors[0].authentifiers = {};
                  unit.authors[0].authentifiers[path] = (0, _internal.sign)(textToSign, privKeyBuf);

                  unit.messages = [].concat(customMessages);
                  unit.unit = (0, _internal.getUnitHash)(unit);

                  return _context.abrupt('return', unit);

                case 62:
                case 'end':
                  return _context.stop();
              }
            }
          }, _callee, this);
        }));

        function message(_x4, _x5) {
          return _ref.apply(this, arguments);
        }

        return message;
      }()
    };

    this.post = {
      message: function () {
        var _ref4 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(app, payload, options) {
          var unit;
          return _regenerator2.default.wrap(function _callee2$(_context2) {
            while (1) {
              switch (_context2.prev = _context2.next) {
                case 0:
                  _context2.next = 2;
                  return self.compose.message(app, payload, options);

                case 2:
                  unit = _context2.sent;
                  return _context2.abrupt('return', self.broadcast(unit));

                case 4:
                case 'end':
                  return _context2.stop();
              }
            }
          }, _callee2, this);
        }));

        function message(_x6, _x7, _x8) {
          return _ref4.apply(this, arguments);
        }

        return message;
      }()
    };

    (0, _assign2.default)(this.api, (0, _internal.mapAPI)(_api2.default, requestAsync));
    (0, _assign2.default)(this.compose, (0, _internal.mapAPI)(_apps2.default, this.compose.message));
    (0, _assign2.default)(this.post, (0, _internal.mapAPI)(_apps2.default, this.post.message));
  }

  (0, _createClass3.default)(Client, [{
    key: 'broadcast',
    value: function () {
      var _ref5 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee3(unit) {
        return _regenerator2.default.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return this.api.postJoint({ unit: unit });

              case 2:
                return _context3.abrupt('return', unit.unit);

              case 3:
              case 'end':
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function broadcast(_x9) {
        return _ref5.apply(this, arguments);
      }

      return broadcast;
    }()
  }, {
    key: 'getCachedWitnesses',
    value: function () {
      var _ref6 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee4() {
        return _regenerator2.default.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                if (!this.cachedWitnesses) {
                  _context4.next = 2;
                  break;
                }

                return _context4.abrupt('return', this.cachedWitnesses);

              case 2:
                _context4.next = 4;
                return this.api.getWitnesses();

              case 4:
                this.cachedWitnesses = _context4.sent;
                return _context4.abrupt('return', this.cachedWitnesses);

              case 6:
              case 'end':
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      function getCachedWitnesses() {
        return _ref6.apply(this, arguments);
      }

      return getCachedWitnesses;
    }()
  }, {
    key: 'subscribe',
    value: function subscribe(cb) {
      this.client.subscribe(cb);
    }
  }, {
    key: 'justsaying',
    value: function justsaying(subject, body) {
      this.client.justsaying(subject, body);
    }
  }, {
    key: 'close',
    value: function close() {
      this.client.close();
    }
  }]);
  return Client;
}();

exports.default = Client;
module.exports = exports.default;
